### ISSUE TITLE
Adapter for SubprocessProtocol

### ISSUE THREAD

**Issue by niklasf** [ronf/asyncssh#188]
It looks like `asyncssh.SSHClientSession` is very similar to `SubprocessProtocol` (although strictly speaking the latter is intended for processes on the same machine, and the corresponding transport provides a pid).

Feature request: An adapter that allows running a `SubprocessProtocol` on a remote server. Does this make sense to you / is this something you would consider adding?

**Comment by niklasf** [comment]
Thank you for the detailed explanation.

My use case is this: I have an [existing](https://python-chess.readthedocs.io/en/latest/engine.html#chess.engine.EngineProtocol) SubprocessProtocol that I can run on my local machine:

```python
transport, protocol = await loop.subprocess_exec(chess.engine.UciProtocol, "my-chess-engine")
# ...
```

Using your comments, I am now able to run it on a remote machine:

```python
async def main():
    async with asyncssh.connect("localhost") as conn:
        channel, protocol = await conn.create_session(lambda: SubprocessProtocolAdapter(chess.engine.UciProtocol(), "my-chess-engine")
        # ...

# perhaps better named SSHClientSubprocessSession
class SubprocessProtocolAdapter(asyncssh.SSHClientSession):
    def __init__(self, protocol):
        self.protocol = protocol

    def connection_made(self, channel):
        self.protocol.connection_made(SubprocessTransportAdapter(channel))

    def data_received(self, data, datatype):
        fd = 2 if datatype == asyncssh.EXTENDED_DATA_STDERR else 1
        self.protocol.pipe_data_received(fd, data.encode("utf-8"))

    def eof_received(self):
        pass  # no equivalent, just close the connection entirely

    def connection_lost(self, exc):
        self.protocol.connection_lost(exc)

    def exit_status_received(self, status):
        self.protocol.process_exited()

    def exit_signal_received(self, signal, core_dumped, msg, lang):
        self.protocol.process_exited()

class SubprocessTransportAdapter:
    def __init__(self, channel):
        self.channel = channel

    def get_pipe_transport(self, fd):
        assert fd == 0
        return self

    def write(self, data):
        self.channel.write(data.decode("utf-8"))

    def writelines(self, list_of_data):
        self.channel.writelines("".join(list_of_data).decode("utf-8"))

    def get_pid(self):
        return None  # no local pid available!

    def get_returncode(self):
        return self.channel.get_exit_status()  # or negative signal id

    # etc ...
```

This works nicely. Done. Now I wonder if a (more robust) version of this is worth adding to asyncssh, or if it is too much of a niche use case.

**Comment by ronf** [comment]
I took a closer look at this over the weekend, and I think I've got something working in commit 3deed71 now in the "develop" branch. You invoke it by calling create_subprocess() on SSHClientConnection, instead of something like create_session() or create_process(). When you do that, the first argument is now a protocol_factory which should return something compatible with asyncio.SubprocessProtocol, and like asyncio.BaseEventLoop.subprocess_exec(), it will return a transport compatible with asyncio.SubprocessTransport and the instance of the protocol returned by the protocol_factory you pass in.

As part of this, I went back and added support in both SSHClientChannel and SSHClientProcess for send_signal() to take numeric signal values, as long as they match one of the "SIG" constants in the signal module. I also added support for return codes which are compatible with subprocess.Popen, where processes that exit due to a signal return the negative of the signal number. If the signal name returned by SSH doesn't match anything on the local system, that gets mapped to -99.

In the process, I added the ability to do the same kind of redirect() calls on these new subprocesses as you could do on SSHProcess, to change what stdin, stdout, and stderr are directed to during the lifetime of the remote process. I also added the ability to optionally specify an encoding if you want to read/write Unicode data instead of bytes, but it defaults to bytes so you don't have to explicitly pass in encoding=None to make it work with an existing SubprocessProtocol class that wants bytes.

This implementation also supports pause_reading/resume_reading and pause_writing/resume_writing to provide flow control.

If you get a chance to play with this, I'd love to hear your feedback!

**Comment by niklasf** [comment]
Thanks for implementing this so quickly. I can confirm that it works very well. (But note that the protocols I am using do not exercise pause/resume reading/writing or any other advanced features).

**Comment by ronf** [comment]
Thanks for the report, and I'm glad to hear it is working for you.

I tested the other features here -- I've confirmed that pause/resume reading works well on the pipe transport objects, delaying data being delivered via pipe_data_received(), and the support for pause/resume writing is actually inherited from SSHProcess, so that has already been well tested. I've also tested the redirect support (also inherited from SSHProcess).

This should be available in the next release, which I hope to have by next weekend. There are a few other features I just checked in that I want to get some feedback on.
